dbms/src/Dictionaries/ComplexKeyCacheDictionary.cpp

```diff
@@ -912,28 +913,32 @@ void ComplexKeyCacheDictionary::freeKey(const StringRef key) const
         keys_pool->free(const_cast<char *>(key.data), key.size);
 }
 
-template <typename Arena>
+template <typename Pool>
 StringRef ComplexKeyCacheDictionary::placeKeysInPool(
-    const size_t row, const ConstColumnPlainPtrs & key_columns, StringRefs & keys, Arena & pool)
+    const size_t row, const ConstColumnPlainPtrs & key_columns, StringRefs & keys, Pool & pool)
 {
     const auto keys_size = key_columns.size();
     size_t sum_keys_size{};
-    for (const auto i : ext::range(0, keys_size))
+
+    Arena arena;
```

Comment:
Temporary Arena for every row is very expensive.
Better to implement overload `IColumn::serializeValueIntoArena` for `ArenaWithFreeLists`.
(or call it `IColumn::serializeValueIntoArenaWithFreeLists` and provide a wrapper here.)